File: kotlin/custom-set/src/main/kotlin/CustomSet.kt
class CustomSet(vararg items: Int) {

    private val items = items.toMutableSet()

    fun isEmpty(): Boolean = items.isEmpty()

    fun isSubset(other: CustomSet) = other.items.containsAll(items)

    fun isDisjoint(other: CustomSet) = intersection(other).isEmpty()

    fun contains(other: Int) = items.contains(other)

    fun intersection(other: CustomSet) = CustomSet(*items.intersect(other.items).toIntArray())

    fun add(other: Int) {
        items += other
    }

    override fun equals(other: Any?): Boolean = when (other) {
        is CustomSet -> items == other.items
        else -> false
    }

    operator fun plus(other: CustomSet) = CustomSet(*(items + other.items).toIntArray())

    operator fun minus(other: CustomSet) = CustomSet(*(items - other.items).toIntArray())

    override fun hashCode(): Int = items.hashCode()
}

File: swift/custom-set/Sources/CustomSet/CustomSet.swift
class CustomSet: Equatable {
  private var items: [Int] = []

  init(_ items: [Int]) {
    items.forEach(add)
  }

  var isEmpty: Bool {
    items.isEmpty
  }

  func add(_ item: Int) {
    if !contains(item) {
      items.append(item)
    }
  }

  func contains(_ item: Int) -> Bool {
    items.contains(item)
  }

  func isSubset(of other: CustomSet) -> Bool {
    items.allSatisfy(other.contains)
  }

  func isDisjoint(with other: CustomSet) -> Bool {
    items.allSatisfy({ !other.contains($0) })
  }

  func intersection(_ other: CustomSet) -> CustomSet {
    CustomSet(items.filter(other.contains))
  }

  func difference(_ other: CustomSet) -> CustomSet {
    CustomSet(items.filter { !other.contains($0) })
  }

  func union(_ other: CustomSet) -> CustomSet {
    CustomSet(items + other.items)
  }

  static func == (lhs: CustomSet, rhs: CustomSet) -> Bool {
    lhs.isSubset(of: rhs) && rhs.isSubset(of: lhs)
  }
}
